txnsync/txngroups.go

```diff
@@ -17,33 +17,202 @@
+	BitmaskSelectionPK      bitmask                           `codec:"selkeybm,allocbound=maxBitmaskSize"`
+	VoteFirst               []basics.Round                    `codec:"votefst,allocbound=maxEncodedTransactionGroup"`
+	BitmaskVoteFirst        bitmask                           `codec:"votefstbm,allocbound=maxBitmaskSize"`
+	VoteLast                []basics.Round                    `codec:"votelst,allocbound=maxEncodedTransactionGroup"`
+	BitmaskVoteLast         bitmask                           `codec:"votelstbm,allocbound=maxBitmaskSize"`
+	VoteKeyDilution         []uint64                          `codec:"votekd,allocbound=maxEncodedTransactionGroup"`
+	BitmaskVoteKeyDilution  bitmask                           `codec:"votekdbm,allocbound=maxBitmaskSize"`
+	Nonparticipation        []bool                            `codec:"nonpart,allocbound=maxEncodedTransactionGroup"`
+	BitmaskNonparticipation bitmask                           `codec:"nonpartbm,allocbound=maxBitmaskSize"`
+}
+
+type encodedPaymentTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Receiver        []basics.Address    `codec:"rcv,allocbound=maxEncodedTransactionGroup"`
+	BitmaskReceiver bitmask             `codec:"rcvbm,allocbound=maxBitmaskSize"`
+	Amount          []basics.MicroAlgos `codec:"amt,allocbound=maxEncodedTransactionGroup"`
+	BitmaskAmount   bitmask             `codec:"amtbm,allocbound=maxBitmaskSize"`
+
+	CloseRemainderTo        []basics.Address `codec:"close,allocbound=maxEncodedTransactionGroup"`
+	BitmaskCloseRemainderTo bitmask          `codec:"closebm,allocbound=maxBitmaskSize"`
+}
+
+type encodedAssetConfigTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+}
+
+type encodedAssetTransferTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+}
+
+type encodedAssetFreezeTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+}
+
+type encodedApplicationCallTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+}
+
+type encodedCompactCertTxnFields struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+}
+
+type bitmask []byte
+
+// assumed to be in mode 0, sets bit at index to 1
+func (b bitmask) SetBit(index int) {
+	byteIndex := index/8 + 1
+	b[byteIndex] |= 1 << (index % 8)
+}
+
+func (b bitmask) EntryExists(index int, hint int) (bool, int) {
+	option := 0
+	if len(b) > 0 {
+		option = int(b[0])
+	}
+	switch option {
+	case 0: // if we have the bit 1 then we have an entry at the corresponding bit index.
+		byteIndex := index/8 + 1
+		return byteIndex < len(b) && (b[byteIndex]&(1<<(index%8)) != 0), 0
+	case 1: // if we have the bit 0 then we have an entry at the corresponding bit index.
+		byteIndex := index/8 + 1
+		return byteIndex >= len(b) || (b[byteIndex]&(1<<(index%8)) == 0), 0
+	case 2: // contains a list of bytes designating the transaction bit index
+		for hint < len(b) && index <= int(b[index]) {
```

Comment:
sounds like a binary search would be more suitable here. but for compactness sake, we should use differential list.

Reply:
In practice calling this method over all the indices will iterate over the entire list just once.

Also, what is a differential list / why would that be helpful?

Reply:
differential list goes as follows:
instead of storing a list of [50, 60, 70, 100], we'll store a list of [50, 10, 10, 30]. While this doesn't seems like a reduction is size, it would allow msgp to use less bytes for encoding these. ( since msgp spends less bytes on integer less than 127 ).
Another benefits is the increase the frequency of "small" numbers. This helps to improve the compression rate, if we were to apply a compression on top of that.